tests/test_estimation/test_calculate_weights.py

```diff
@@ -0,0 +1,454 @@
+        [2, np.nan, "P", "210", "0006", 0, 1, False],
+        [3, 1, np.nan, "210", "0006", 0, 1, False],
+        [4, 1, "P", np.nan, "0006", 0, 1, False],
+        [5, 1, "P", "210", np.nan, 0, 1, False],
+        [6, 1, "P", "210", "0006", np.nan, 2, False],
+        [7, 1, "P", "210", "0006", 0, np.nan, False],
+        [8, 1, "P", "210", "0006", 0, 2, np.nan],
+        [9, 1, "P", "210", "0006", 0, 2, False],
+    ]
+
+    input_df = pd.DataFrame(data=data, columns=input_cols)
+
+    cellno_dict = {1: 10, 2: 5}
+
+    expected_cols = [
+        "reference",
+        "709",
+        "selectiontype",
+        "statusencoded",
+        "formtype",
+        "instance",
+        "cellnumber",
+        "outlier",
+        "a_weight",
+    ]
+
+    data = [
+        [1, 1, "P", "210", "0006", 0, 1, False, 10.0],
+        [2, np.nan, "P", "210", "0006", 0, 1, False, 10.0], # filtered from calc but weight applied
+        [3, 1, np.nan, "210", "0006", 0, 1, False, 1.0],  # filtered out (selectiontype)
+        [4, 1, "P", np.nan, "0006", 0, 1, False, 1.0],    # filtered out (statusencoded)
+        [5, 1, "P", "210", np.nan, 0, 1, False, 1.0],     # filtered out (formtype)
+        [6, 1, "P", "210", "0006", np.nan, 2, False, 2.5],  # filtered out (instance) but weight applied
+        [7, 1, "P", "210", "0006", 0, np.nan, False, 1.0],  # No cellno
+        [8, 1, "P", "210", "0006", 0, 2, np.nan, 2.5],      # No outlier
+        [9, 1, "P", "210", "0006", 0, 2, False, 2.5],
+    ]
+
+    expected_df = pd.DataFrame(data=data, columns=expected_cols)
+
+    expected_qa_cols = [
+        "cellnumber",
+        "N",
+        "n",
+        "outliers",
+        "a_weight",
+    ]
+
+    data = [
+        [1.0, 10.0, 1.0, 0.0, 10.0],
+        [2.0, 5.0, 2.0, 0.0, 2.5],
+    ]
+
+    expected_qa_df = pd.DataFrame(data=data, columns=expected_qa_cols)
+
+    result_df, result_qa_df = calw.calculate_weighting_factor(input_df, cellno_dict)
+
+    assert_frame_equal(result_df, expected_df, check_exact=False, rtol=0.01)
+    assert_frame_equal(result_qa_df, expected_qa_df, check_exact=False, rtol=0.01)
+
+
+# One tests for outlier_weights:
+# test that all appropriate rows are given an a_weight = 1.0
+
+def test_outlier_weights():
+    """Test for outlier_weights."""
+
+    input_cols = [
+        "reference",
+        "outlier",
+    ]
+
+    data = [
+        [1, True],
+        [2, False],
+        [2, True],
+        [4, True],
+        [1, False],
+    ]
+
+    input_df = pd.DataFrame(data=data, columns=input_cols)
+
+    result_df = calw.outlier_weights(input_df)
+
+    expected_cols = [
+        "reference",
+        "outlier",
+        "a_weight",
+    ]
+
+    data = [
+        [1, True, 1.0],
+        [2, False, None],
+        [2, True, 1.0],
+        [4, True, 1.0],
+        [1, False, None],
+    ]
+
+    expected_df = pd.DataFrame(data=data, columns=expected_cols)
+
+    assert_frame_equal(result_df, expected_df)
```

Comment:
if you wanted to save lines here, you could just assert the series for a_weight. This doesn't make a difference to the actual test however so just optional :).
Also, it might not be useful in the context of the functions use, however it appears that there are no defences for cases where the outlier column isn't present